cpp/src_prims/sparse/distance/lp_distance.cuh

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <limits.h>
+#include <raft/cudart_utils.h>
+#include <sparse/distance/common.h>
+#include <sparse/distance/operators.cuh>
+
+#include <raft/cudart_utils.h>
+#include <raft/linalg/distance_type.h>
+#include <raft/sparse/cusparse_wrappers.h>
+#include <raft/cuda_utils.cuh>
+
+#include <common/device_buffer.hpp>
+
+#include <sparse/utils.h>
+#include <sparse/csr.cuh>
+
+#include <sparse/distance/common.h>
+#include <sparse/distance/csr_spmv.cuh>
+
+#include <cuml/common/cuml_allocator.hpp>
+
+#include <nvfunctional>
+
+#include <cusparse_v2.h>
+
+namespace MLCommon {
+namespace Sparse {
+namespace Distance {
+
+template <typename value_idx = int, typename value_t = float, int tpb = 1024,
+          typename reduce_f, typename accum_f, typename write_f>
+
+void unexpanded_lp_distances(
+  value_t *out_dists, const distances_config_t<value_idx, value_t> &config_,
+  reduce_f reduce_func, accum_f accum_func, write_f write_func) {
+  /**
+ * @TODO: Main logic here:
+ *
+ *  - if n_cols < available smem, just use dense conversion for rows of A
+ *  - if n_cols > available smem but max nnz < available smem, use hashing
+ *    (not yet available)
+ *  - if n_cols > available smem & max_nnz > available smem,
+ *              use batching + hashing only for those large cols
+ *  Ref: https://github.com/rapidsai/cuml/issues/3371
+ */
+
+  if (config_.a_ncols < max_cols_per_block<value_idx, value_t, tpb>()) {
+    // TODO: Use n_cols to set shared memory and threads per block
+    // for max occupancy.
+    // Ref: https://github.com/rapidsai/cuml/issues/3371
+
+    raft::mr::device::buffer<value_idx> coo_rows(
+      config_.allocator, config_.stream, max(config_.b_nnz, config_.a_nnz));
```

Comment:
So, this wide path will use a lot of temp memory potentially because it has to make a COO copy, right? Maybe a log debug at least for that would be handy

Reply:
It doesn't need to copy anything, just generate the coo row array (size `nnz`). Sure, it's extra memory, but its at the cost of being able to load balance and it's still much less memory than is being used by the cusparse approach.

Reply:
Also worth noting is that this is just generating a coo row array for the input A for each batch, which is a huge optimization over the current cusparse approach, which needs to transpose B (so a full copy of the CSR) AND allocates a dense array for the output in addition to the sparse CSR  array that’s created for the (already pretty much dense) n^2 output. 

Just for computing the inner products, the savings with this code path is immense. But the cusparse spgemm doesn’t allow us to customize the reduction and accumulation ops so for those distances that require implicit unions instead of intersections, this also is necessary for correctness.